File: server/src/worker/import/parser.ts
import EventEmitter from 'events';
import fs, { ReadStream, createReadStream } from "fs-extra";
import mergeStream from "merge-stream";
import csvparse from "csv-parse";
import { Readable, Transform } from 'stream';
import { Importer } from './importer';
import logger from './logger';
import { AccountingRecord, PaymentRecord, EventRecord, ProfileRecord, YearRecord } from "../../schema/database";
import { ImportRecord } from '../../schema/database/import';
import { paramCase } from 'change-case';

const headerAliases = {
  type: ["type", "recordType", "MODUL", "DOKLAD_AGENDA"],
  paragraph: ["paragraph", "PARAGRAF"],
  item: ["item", "POLOZKA"],
  unit: ["unit", "ORJ"],
  event: ["eventId", "event", "AKCE", "ORG"],
  amount: ["amount", "CASTKA"],
  date: ["date", "DATUM", "DOKLAD_DATUM"],
  counterpartyId: ["counterpartyId", "SUBJEKT_IC"],
  counterpartyName: ["counterpartyName", "SUBJEKT_NAZEV"],
  description: ["description", "POZNAMKA"],
  id: ["id", "eventId", "srcId", "AKCE", "ORG"],
  name: ["name", "eventName", "AKCE_NAZEV"]
};

const mandatoryAccountingHeaders = [
  "type",
  "paragraph",
  "item",
  "event",
  "amount"
]

const mandatoryPaymentsHeaders = [
  "type",
  "paragraph",
  "item",
  "unit",
  "event",
  "amount",
  "date",
  "counterpartyId",
  "counterpartyName",
  "description"
]

const mandatoryEventHeaders = [
  "id",
  "name"
]

export class ImportParser {

  static createCsvParser(file: string): csvparse.Parser {
    let headers = []
    switch (file) {
      case "accountingFile":
        headers = mandatoryAccountingHeaders
        break
      case "paymentsFile":
      case "dataFile":
        headers = mandatoryPaymentsHeaders
        break
      case "eventsFile":
        headers = mandatoryEventHeaders
        break
      default:
        throw Error("Unexpected parser requested")
    }

    return csvparse({ delimiter: ";", columns: line => this.parseHeader(line, headers), relax_column_count: true });
  }

  static createCsvTransformer(file: string, options: Importer.Options): Transform {
    switch (file) {
      case "paymentsFile":
      case "dataFile":
      case "accountingFile":
        return this.createDataTransformer(options)
      case "eventsFile":
        return this.createEventsTransformer(options)
      default:
        throw Error("Unexpected parser requested")
    }
  }

  static createDataTransformer(options: Importer.Options) {
    return new Transform({
      writableObjectMode: true,
      readableObjectMode: true,
      transform: function (line, enc, callback) {
        const recordType = line["type"]

        if (recordType === "KDF" || recordType === "KOF") {
          try {
            let payment = ImportParser.createPaymentRecord(line, options)
            this.push({type: "payment", record: payment});
            callback()
          } catch (err) {
            callback(err)  
          }
        } else {
          var accounting = ImportParser.createAccountingRecord(line, options)
          this.push({type: "accounting", record: accounting});
          callback()
        } 
      }
    });

  }

  static createEventsTransformer(options: Importer.Options) {
    return new Transform({
      writableObjectMode: true,
      readableObjectMode: true,
      transform: function (line, enc, callback) {
        let event = ImportParser.createEventRecord(line, options)
        if (line.id && line.name) {
          this.push({type: "event", record: event});
          callback()
        }
        else {
          callback(new Error(`Missing event id; event name: ${line.name}`))
        }
    }
    });

  }

  static parseHeader(headerLine: string[], headerNames: string[]): string[] {
    // remove possible BOM at the beginning of file, also removes extra whitespaces
    headerLine = headerLine.map(item => item.trim())
    logger.log(`Searching for these headers: [${headerNames}]`)
    logger.log(`The header array being searched for field names: [${headerLine}]`)
    const foundHeaders: string[] = headerLine.map(originalField => {
      // browse through all the target fields if originalField is someones alias
      return Object.keys(headerAliases).find(key => headerAliases[key].indexOf(originalField) != -1)
    });
    headerNames.forEach(h => {
      if (foundHeaders.indexOf(h) == -1) {
        throw Error(`Failed to find column header "${h}"`)
      }
    })

    return foundHeaders

  }

  static createPaymentRecord(row: {}, options: Importer.Options): PaymentRecord {
    return ["type", "paragraph", "item", "event", "amount", "date", "counterpartyId", "counterpartyName", "description"].reduce((acc, c) => {
      switch(c) {
        case "date":
          const d = Date.parse(row[c])
          if (isNaN(d) || !/\d{4}-\d{2}-\d{2}/.test(row[c])) this.invalidField("date", "date", row)
          acc[c] = row[c]
          break
        case "paragraph":
        case "item":
        case "unit":
        case "event":
        case "amount":
          const n = Number(row[c])
          if (isNaN(n)) this.invalidField(c, "number", row) 
          acc[c] = n
          break
        case "counterpartyId":
        case "counterpartyName":
        case "description":
          acc[c] = row[c]
          break
      }
      return acc
    }, {
      profileId: options.profileId,
      year: options.year
    } as PaymentRecord)
  }

  static createAccountingRecord(row: {}, options: Importer.Options): AccountingRecord {
    return ["type", "paragraph", "item", "event", "unit", "amount"].reduce((acc, c) => {
      switch(c) {
        case "type":
          acc[c] = row[c]
          break
        case "paragraph":
        case "item":
        case "event":
        case "unit":
        case "amount":
          const n = Number(row[c])
          if (isNaN(n)) this.invalidField(c, "number", row) 
          acc[c] = n
          break
      }
      return acc
    }, {
      profileId: options.profileId,
      year: options.year
    } as AccountingRecord)
  }

  static createEventRecord(row: {}, options: Importer.Options): EventRecord {
    return ["id", "name", "description"].reduce((acc, c) => {
      switch(c) {
        case "name":
        case "description":
          acc[c] = row[c]
          break
        case "id":
          const n = Number(row[c])
          if (isNaN(n)) this.invalidField(c, "number", row) 
          acc[c] = n
          break
      }
      return acc
    }, {
      profileId: options.profileId,
      year: options.year
    } as EventRecord)
  }

  private static invalidField(field: string, type: string, row: {}): never {
    throw new Error(`Failed to convert field "${field}": ${row[field]} to ${type}.\nRow processed: ${JSON.stringify(row)}`)
  }

}

File: validator/validator/lib/parser.py
import csv
from time import strptime
from typing import TextIO, Iterable, List, Dict, Union

from .message import Message, Level


class Header:
    aliases: List[str] = []
    _real: str = ''

    def __init__(self, aliases: List[str]) -> None:
        super().__init__()
        self.aliases = aliases

    @property
    def real(self):
        return self._real if self._real != '' else self.aliases[0]

    @real.setter
    def real(self, value):
        self._real = value


class Parser:
    headers: Dict[str, Header] = {}

    @staticmethod
    def create(input_file: str) -> Union['Parser', None]:
        if input_file.endswith('_events.csv'):
            return EventsFileParser()
        if input_file.endswith('_data.csv'):
            return DataFileParser()
        return None

    def parse(self, input: TextIO) -> Iterable[Message]:
        reader = csv.DictReader(input, delimiter=';')
        errors: List[Message] = []

        self.__populate_headers(reader.fieldnames)

        errors += self.__validate_header_keys(reader.fieldnames)
        if any(m.level is Level.ERROR for m in errors):
            return errors

        for position, row in enumerate(reader):
            errors += self._validate_row(row, position + 1)

        return errors

    def _required_headers(self):
        return []

    def __populate_headers(self, keys: Iterable[str]):
        for key in keys:
            for header_key in self.headers.keys():
                if key in self.headers[header_key].aliases:
                    self.headers[header_key].real = key

    def __validate_header_keys(self, keys: Iterable[str]) -> Iterable[Message]:
        errors = []
        for key in keys:
            if key not in [f.real for f in self.headers.values()]:
                errors.append(Message(Level.WARNING, 0, 'Unknown key: "{}". It will be ignored.'.format(key)))
        for key in [self.headers[header].real for header in self._required_headers()]:
            if key not in keys:
                errors.append(Message(Level.ERROR, 0, 'Missing required key "{}".'.format(key)))
        return errors

    def _validate_row(self, row: Dict[str, str], line: int) -> Iterable[Message]:
        return []


class EventsFileParser(Parser):
    headers = {
        "id": Header(["id", "eventId", "srcId", "AKCE", "ORG"]),
        "name": Header(["name", "eventName", "AKCE_NAZEV"]),
        "description": Header(["description"]),
    }

    def _required_headers(self):
        return ['id', 'name']

    def _validate_row(self, row: Dict[str, str], line: int) -> Iterable[Message]:
        return super()._validate_row(row, line)


class DataFileParser(Parser):
    headers = {
        'type': Header(["type", "recordType", "MODUL", "DOKLAD_AGENDA"]),
        'paragraph': Header(["paragraph", "PARAGRAF"]),
        'item': Header(["item", "POLOZKA"]),
        'unit': Header(["unit", "ORJ"]),
        'event': Header(["eventId", "event", "AKCE", "ORG"]),
        'amount': Header(["amount", "CASTKA"]),
        'date': Header(["date", "DATUM", "DOKLAD_DATUM"]),
        'counterpartyId': Header(["counterpartyId", "SUBJEKT_IC"]),
        'counterpartyName': Header(["counterpartyName", "SUBJEKT_NAZEV"]),
        'description': Header(["description", "POZNAMKA"]),
    }

    def _required_headers(self):
        return ['type', 'paragraph', 'item', 'amount']

    def _validate_row(self, row: Dict[str, str], line: int) -> Iterable[Message]:
        errors: List[Message] = []
        data_type = row[self.headers['type'].real]
        data_counterparty_id = row[self.headers['counterpartyId'].real]
        data_counterparty_name = row[self.headers['counterpartyName'].real]
        data_description = row[self.headers['description'].real]

        if data_type not in [TYPE_KDF, TYPE_KOF]:
            if data_counterparty_id != '':
                errors.append(self.__type_warning(line, data_type, self.headers['counterpartyId'].real))
            if data_counterparty_name != '':
                errors.append(self.__type_warning(line, data_type, self.headers['counterpartyName'].real))
            if data_description != '':
                errors.append(self.__type_warning(line, data_type, self.headers['description'].real))

        errors += validate_int(line, self.headers['paragraph'].real, row[self.headers['paragraph'].real])
        errors += validate_int(line, self.headers['item'].real, row[self.headers['item'].real])
        errors += validate_empty_or_int(line, self.headers['event'].real, row[self.headers['event'].real])
        errors += validate_empty_or_int(line, self.headers['unit'].real, row[self.headers['unit'].real])
        errors += validate_float(line, self.headers['amount'].real, row[self.headers['amount'].real])
        errors += validate_empty_or_date(line, self.headers['date'].real, row[self.headers['date'].real])

        return errors

    def __type_warning(self, line: int, data_type: str, field: str) -> Message:
        return Message(
            Level.WARNING,
            line,
            '"{}" is processed for types "KDF" or "KOF" only. '
            'Type is "{}" so it will be skipped.'.format(field, data_type),
        )


TYPE_KDF = 'KDF'
TYPE_KOF = 'KOF'


def validate_int(line: int, key: str, value: str) -> Iterable[Message]:
    try:
        int(value)
        return []
    except ValueError:
        return [Message(
            Level.ERROR,
            line,
            '{} expected to be integer number. Actual value: "{}".'.format(key, value)
        )]


def validate_empty_or_int(line: int, key: str, value: str) -> Iterable[Message]:
    if value is None or value == '':
        return []
    return validate_int(line, key, value)


def validate_float(line: int, key: str, value: str) -> Iterable[Message]:
    try:
        float(value)
        return []
    except ValueError:
        return [Message(
            Level.ERROR,
            line,
            '{} expected to be float number. Actual value: "{}".'.format(key, value)
        )]


def validate_empty_or_date(line: int, key: str, value: str) -> Iterable[Message]:
    if value is None or value == '':
        return []
    try:
        strptime(value, '%Y-%m-%d')
        return []
    except ValueError:
        return [Message(
            Level.ERROR,
            line,
            '{} expected to be date in format "YYYY-MM-DD". Actual value: "{}".'.format(key, value)
        )]
